src/py42/clients/settings/__init__.py

```diff
@@ -0,0 +1,199 @@
+from py42._compat import string_type
+from py42.clients.settings._converters import bool_to_str
+from py42.clients.settings._converters import no_conversion
+from py42.clients.settings._converters import str_to_bool
+
+
+def set_val(d, keys, value):
+    """Helper for setting nested values from a dict based on a list of keys."""
+    d = get_val(d, keys[:-1])
+    d[keys[-1]] = value
+
+
+def get_val(d, keys):
+    """Helper for getting nested values from a dict based on a list of keys."""
+    for key in keys:
+        d = d[key]
+    return d
+
+
+def show_change(val1, val2):
+    if isinstance(val1, string_type):
+        val1 = '"{}"'.format(val1)
+    if isinstance(val2, string_type):
+        val2 = '"{}"'.format(val2)
+    return "{} -> {}".format(val1, val2)
+
+
+class BaseSettingProperty(object):
+    def __init__(self, name, location):
+        self.name = name
+        self.location = location
+        self.init_val = None
+
+    def _register_change(self, instance, orig_val, new_val):
+        name = self.name.lstrip("_")
+        if self.init_val is None:
+            self.init_val = orig_val
+        if self.init_val == new_val:
+            if name in instance.changes:
+                instance.changes.pop(name)
+        else:
+            instance.changes[name] = show_change(self.init_val, new_val)
+
+
+class SettingProperty(BaseSettingProperty):
+    """Descriptor class to help manage changes to nested dict values. Assumes attributes
+    being managed are on a UserDict/UserList subclass.
+
+    Args:
+        name (str): name of attribute this class manages (changes will be registered with this name).
+        location (list): list of keys defining the location of the value being managed in the managed class.
+        get_converter (func, optional): function to convert retrieved values to preferred format. Defaults to no conversion.
+        set_converter (func, optional): function to convert values being set to preferred format. Defaults to no conversion.
+    """
+
+    def __init__(
+        self, name, location, get_converter=no_conversion, set_converter=no_conversion
```

Comment:
Instead of `no_conversion()`, since it is only used as a default, these could just be empty lambdas... `..., get_converter=lambda x: x, set_converter=lambda x: x)`

Reply:
Maybe I'll just set the default to `None` and then only call the converter if it's not `None`.